### PR TITLE
Untangling lifetimes and borrows

### PR DESCRIPTION
**Pull request by richardpringle**
- Don't cast lifetimes
- Stop passing in a part of self as a separate arg
- Untangling lifetimes and borrows

It's a miracle that this compiled...

### CODE REVIEW

File: shale/src/compact.rs [hunk 1]
```diff
@@ -654,14 +655,16 @@ impl<T: Storable + Send + Sync + Debug + 'static, M: CachedStore + Send + Sync>
...
 
-        Ok(inner.obj_cache.put(inner.get_data_ref(ptr, payload_size)?))
+        Ok(ObjRef::new(Some(obj), cache))
     }
 
     fn flush_dirty(&self) -> Option<()> {
         let mut inner = self.inner.write().unwrap();
         inner.header.flush_dirty();
-        inner.obj_cache.flush_dirty()
+        self.obj_cache.flush_dirty()
```
Comment (rkuris) [inline shale/src/compact.rs:667]: The guard now lives too long. Should be:
```
self.inner.write().unwrap().header.flush_dirty()
self.obj_cache.flush_dirty()
```

File: shale/src/lib.rs [hunk 2]
```diff
@@ -190,17 +189,12 @@ impl<T: ?Sized + Send + Sync> Deref for Obj<T> {
 /// User handle that offers read & write access to the stored [ShaleStore] item.
 pub struct ObjRef<'a, T: Send + Sync> {
     inner: Option<Obj<T>>,
-    cache: ObjCache<T>,
-    _life: PhantomData<&'a mut ()>,
```
Comment (rkuris) [inline shale/src/lib.rs:194]: 🏆

**Review by rkuris — APPROVED** [review]
Please fix the guard issue before landing

**Review by xinifinity — APPROVED** [review]

File: shale/src/compact.rs (hunk 1 above)
Comment (richardpringle) [inline shale/src/compact.rs:667]:
```suggestion
        // hold the write lock to ensure that both cache and header are flushed in-sync
        self.obj_cache.flush_dirty()
```

Comment (richardpringle) [reply shale/src/compact.rs]: @rkuris, I added a comment here to keep these actions in-sync. Let me know if you still think I should change this

**Pull request merged**